src/append_session.rs

```diff
@@ -196,81 +223,104 @@ where
         inflight,
         inflight_size,
         request_stream,
-        last_acked_seqnum,
+        total_records,
+        total_records_acknowledged,
+        stashed_request,
     } = lock.deref_mut();
 
     assert!(*inflight_size <= stream_client.inner.config.max_append_inflight_bytes);
-    let (input_tx, mut ack_stream) = connect(&stream_client, frame_signal.clone()).await?;
+
+    let (s2_input_tx, mut s2_ack_stream) = connect(&stream_client, frame_signal.clone()).await?;
     let batch_ack_deadline = stream_client.inner.config.request_timeout;
 
     if !inflight.is_empty() {
         resend(
             batch_ack_deadline,
             inflight,
             inflight_size,
-            input_tx.clone(),
-            &mut ack_stream,
-            last_acked_seqnum,
+            s2_input_tx.clone(),
+            &mut s2_ack_stream,
+            total_records_acknowledged,
             output_tx.clone(),
         )
         .await?;
 
-        assert_eq!(inflight.len(), 0);
-        assert_eq!(*inflight_size, 0);
         frame_signal.reset();
 
-        debug!("finished resending");
+        assert!(inflight.is_empty());
+        assert_eq!(*inflight_size, 0);
+        assert_eq!(total_records, total_records_acknowledged);
     }
+
     let timer = MuxTimer::<N_TIMER_VARIANTS>::default();
     tokio::pin!(timer);
-    let mut input_terminated = false;
 
-    while !(input_terminated && inflight.is_empty()) {
+    let mut client_input_terminated = false;
+    let mut stashed_ack = None;
+
+    while !(client_input_terminated && inflight.is_empty()) {
         tokio::select! {
-            (event_ord, _deadline) = &mut timer,
-                if timer.is_armed()
-            => {
+            (event_ord, _deadline) = &mut timer, if timer.is_armed() => {
                 match TimerEvent::from(event_ord) {
-                    TimerEvent::MetricUpdate => {
-                        todo!()
-                    }
-                    TimerEvent::BatchDeadline =>
-                        Err(ClientError::Service(Status::cancelled("sdk hit deadline (`request_timeout`) waiting for an append acknowledgement")))?
+                    TimerEvent::MetricUpdate => todo!(),
+                    TimerEvent::BatchDeadline => Err(ClientError::Service(Status::cancelled("sdk hit local deadline (`request_timeout`) waiting for an append acknowledgement")))?
                 }
             }
-            client_input = request_stream.next(),
-                if !input_terminated && *inflight_size + MAX_BATCH_SIZE <= stream_client.inner.config.max_append_inflight_bytes
+
+            next_request = request_stream.next(), if
+                stashed_request.is_none() &&
+                !client_input_terminated &&
+                *inflight_size + MAX_BATCH_SIZE <= stream_client.inner.config.max_append_inflight_bytes
```

Comment:
unrelated to this PR but we should clarify on the doc for this config that it is applied _per append session_